central/reports/scheduler/schedule.go

```diff
@@ -411,6 +439,91 @@ func (s *scheduler) runPaginatedQuery(ctx context.Context, scopeQuery, cveQuery
 	return resultData, nil
 }
 
+// Returns vuln report data from deployments matched by embedded resource collection.
+func (s *scheduler) runPaginatedDeploymentsQuery(ctx context.Context, cveQuery string, deploymentIds []string) (common.Result, error) {
+	offset := paginatedQueryStartOffset
+	var resultData common.Result
+	for {
+		if offset >= len(deploymentIds) {
+			break
+		}
+		// deploymentsQuery is a *v1.Query and graphQL resolvers accept a string queries.
+		// With nested collections, deploymentsQuery could be a disjunction of conjunctions like the example below.
+		// [(Cluster: c1 AND Namespace: n1 AND Deployment: d1) OR (Cluster: c2 AND Namespace: n2 AND Deployment: d2)]
+		// Current string query language doesn't support disjunction of multiple conjunctions, so we cannot build a
+		// string equivalent of deploymentsQuery for graphQL. Because of this, we first fetch deploymentIDs from
+		// deploymentDatastore using deploymentsQuery and then build string query for graphQL using those deploymentIDs.
+		scopeQuery := fmt.Sprintf("%s:%s", search.DeploymentID.String(),
+			strings.Join(deploymentIds[offset:mathutil.MinInt(offset+numDeploymentsLimit, len(deploymentIds))], ","))
+		r, err := s.execReportDataQuery(ctx, reportQueryPostgres, scopeQuery, cveQuery, paginatedQueryStartOffset)
+		if err != nil {
+			return r, err
+		}
+		resultData.Deployments = append(resultData.Deployments, r.Deployments...)
+		offset += numDeploymentsLimit
+	}
+	return resultData, nil
+}
+
+func (s *scheduler) execReportDataQuery(ctx context.Context, gqlQuery, scopeQuery, cveQuery string, offset int) (common.Result, error) {
+	response := s.Schema.Exec(ctx,
+		gqlQuery, "getVulnReportData", map[string]interface{}{
+			"scopequery": scopeQuery,
+			"cvequery":   cveQuery,
+			"pagination": map[string]interface{}{
+				"offset": offset,
+				"limit":  numDeploymentsLimit,
+				"sortOption": map[string]interface{}{
+					"field":    deploymentSortOption.GetField(),
+					"reversed": deploymentSortOption.GetReversed(),
+				},
+			},
+		})
+	if len(response.Errors) > 0 {
+		log.Errorf("error running graphql query: %s", response.Errors[0].Message)
+		return common.Result{}, response.Errors[0].Err
+	}
+	var res common.Result
+	if err := json.Unmarshal(response.Data, &res); err != nil {
+		return common.Result{}, err
+	}
+	return res, nil
+}
+
+func (s *scheduler) getDeploymentIDs(ctx context.Context, deploymentsQuery *v1.Query) ([]string, error) {
+	deploymentsQuery.Pagination = &v1.QueryPagination{
+		SortOptions: []*v1.QuerySortOption{
+			deploymentSortOption,
+		},
+	}
+	results, err := s.deploymentDatastore.Search(ctx, deploymentsQuery)
+	if err != nil {
+		return nil, err
+	}
+	return search.ResultsToIDs(results), nil
+}
+
+func groupByClusterAndNamespace(result common.Result) common.Result {
```

Comment:
Could you pass the pointer to the constructed result set instead of creating copy?